In [257]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import seaborn as sns
import matplotlib.pyplot as plt




https://www.folkhalsomyndigheten.se/the-public-health-agency-of-sweden/living-conditions-and-lifestyle/andtg/tobacco/towards-a-smoke-free-sweden/

In [258]:
data_sw = {
    1980: 31.0,
    1982: 30.5,
    1984: 28.5,
    1986: 27.0,
    1988: 26.0,
    1990: 24.5,
    1992: 23.0,
    1994: 22.0,
    1996: 20.0,
    1998: 18.0,
    2000: 17.5,
    2002: 16.5,
    2004: 15.0,
    2006: 14.0,
    2007: 13.8,
    2008: 13.0,
    2010: 12.5,
    2012: 11.8,
    2014: 11.0,
    2016: 10.5,
    2018: 10.0,
    2020: 9.8,
    2021: 9.7,
    2022: 9.5,
    2023: 8.8
}

https://thl.fi/documents/155392151/190513223/TobaccoStatistics2022_front-page+appendix-tables.pdf/bf593cfb-4d73-2c7d-d4a9-c2025d4a372d/TobaccoStatistics2022_front-page+appendix-tables.pdf?t=1698215141348

In [259]:
data_fin = {
    1997: 19,
    1999: 18,
    2001: 18,
    2003: 16,
    2005: 16,
    2007: 15,
    2009: 14,
    2011: 13,
    2013: 11,
    2014: 14,
    2015: 14,
    2016: 13,
    2017: 11,
    2018: 12,
    2019: 11,
    2020: 11,
    2022: 10
}

In [260]:
df_den = pd.read_excel("graveyard/roykere_danmark.xlsx").drop(columns=["Unnamed: 0"])

# Strip whitespace from column names to ensure the rename works
df_den.columns = df_den.columns.str.strip()

df_den = df_den.rename(columns={"Percentage" : "smokers_percentage_den"})
df_den

Year  smokers_percentage_den
0   1998                      35
1   1999                      34
2   2000                      33
3   2001                      32
4   2002                      31
5   2003                      30
6   2004                      27
7   2005                      28
8   2006                      28
9   2007                      28
10  2008                      28
11  2009                      24
12  2010                      23
13  2011                      22
14  2012                      23
15  2013                      22
16  2014                      21
17  2015                      22
18  2016                      21
19  2017                      22
20  2018                      23
21  2019                      20
22  2020                      18
23  2021                      19
24  2022                      19
25  2023                      18
26  2024                      17

In [261]:
df_no = pd.read_excel("data_washed/norway_smokers.xlsx").drop(columns=["Unnamed: 0"])
df_no = df_no.rename(columns={"Unnamed: 2": "Year", "Unnamed: 3": "smokers_percentage_no"})
df_no["Year"] = pd.to_numeric(df_no["Year"], errors="coerce")
df_no = df_no.dropna(subset=["Year", "smokers_percentage_no"]).reset_index(drop=True)
df_no["Year"] = df_no["Year"].astype(int)
df_no

Year  smokers_percentage_no
0   1973                   42.0
1   1974                   41.0
2   1975                   41.0
3   1976                   39.0
4   1977                   38.0
5   1978                   38.0
6   1979                   37.0
7   1980                   36.0
8   1981                   36.0
9   1982                   36.0
10  1983                   37.0
11  1984                   37.0
12  1985                   37.0
13  1986                   36.0
14  1987                   37.0
15  1988                   37.0
16  1989                   36.0
17  1990                   35.0
18  1991                   35.0
19  1992                   34.0
20  1993                   36.0
21  1994                   33.0
22  1995                   33.0
23  1996                   33.0
24  1997                   34.0
25  1998                   33.0
26  1999                   31.0
27  2000                   31.0
28  2001                   29.0
29  2002                   29.0
30  2003                   26.0
31  2004                   26.0
32  2005                   25.0
33  2006                   23.0
34  2007                   21.0
35  2008                   20.0
36  2009                   20.0
37  2010                   19.0
38  2011                   17.0
39  2012                   16.0
40  2013                   15.0
41  2014                   13.0
42  2015                   13.0
43  2016                   12.0
44  2017                   11.0
45  2018                   11.0
46  2019                    9.0
47  2020                    9.0
48  2021                    8.0
49  2022                    8.0
50  2023                    7.0
51  2024                    8.0

In [262]:
df_sw = pd.Series(data_sw, name='smokers_percentage_sw').reset_index().rename(columns={'index': 'Year'})
df_sw['Year'] = pd.to_numeric(df_sw['Year'], errors='coerce').astype(int)

df_fin = pd.Series(data_fin, name='smokers_percentage_fin').reset_index().rename(columns={'index': 'Year'})
df_fin['Year'] = pd.to_numeric(df_fin['Year'], errors='coerce').astype(int)

In [263]:
from functools import reduce

# List of all dataframes to merge
data_frames = [df_no, df_sw, df_den, df_fin]

# Use reduce to sequentially merge all dataframes in the list on 'Year' with an outer join
df = reduce(lambda left, right: pd.merge(left, right, on='Year', how='outer'), data_frames)

# Sort by year, which is good practice after an outer merge
df = df.sort_values('Year').reset_index(drop=True)

df

Year  smokers_percentage_no  smokers_percentage_sw  \
0   1973                   42.0                    NaN   
1   1974                   41.0                    NaN   
2   1975                   41.0                    NaN   
3   1976                   39.0                    NaN   
4   1977                   38.0                    NaN   
5   1978                   38.0                    NaN   
6   1979                   37.0                    NaN   
7   1980                   36.0                   31.0   
8   1981                   36.0                    NaN   
9   1982                   36.0                   30.5   
10  1983                   37.0                    NaN   
11  1984                   37.0                   28.5   
12  1985                   37.0                    NaN   
13  1986                   36.0                   27.0   
14  1987                   37.0                    NaN   
15  1988                   37.0                   26.0   
16  1989                   36.0                    NaN   
17  1990                   35.0                   24.5   
18  1991                   35.0                    NaN   
19  1992                   34.0                   23.0   
20  1993                   36.0                    NaN   
21  1994                   33.0                   22.0   
22  1995                   33.0                    NaN   
23  1996                   33.0                   20.0   
24  1997                   34.0                    NaN   
25  1998                   33.0                   18.0   
26  1999                   31.0                    NaN   
27  2000                   31.0                   17.5   
28  2001                   29.0                    NaN   
29  2002                   29.0                   16.5   
30  2003                   26.0                    NaN   
31  2004                   26.0                   15.0   
32  2005                   25.0                    NaN   
33  2006                   23.0                   14.0   
34  2007                   21.0                   13.8   
35  2008                   20.0                   13.0   
36  2009                   20.0                    NaN   
37  2010                   19.0                   12.5   
38  2011                   17.0                    NaN   
39  2012                   16.0                   11.8   
40  2013                   15.0                    NaN   
41  2014                   13.0                   11.0   
42  2015                   13.0                    NaN   
43  2016                   12.0                   10.5   
44  2017                   11.0                    NaN   
45  2018                   11.0                   10.0   
46  2019                    9.0                    NaN   
47  2020                    9.0                    9.8   
48  2021                    8.0                    9.7   
49  2022                    8.0                    9.5   
50  2023                    7.0                    8.8   
51  2024                    8.0                    NaN   

    smokers_percentage_den  smokers_percentage_fin  
0                      NaN                     NaN  
1                      NaN                     NaN  
2                      NaN                     NaN  
3                      NaN                     NaN  
4                      NaN                     NaN  
5                      NaN                     NaN  
6                      NaN                     NaN  
7                      NaN                     NaN  
8                      NaN                     NaN  
9                      NaN                     NaN  
10                     NaN                     NaN  
11                     NaN                     NaN  
12                     NaN                     NaN  
13                     NaN                     NaN  
14                     NaN                     NaN  
15                     NaN                     NaN  
16                     NaN  

In [264]:
df = df.sort_values('Year').reset_index(drop=True)

# Create a complete range of years from min to max
full_years = pd.DataFrame({'Year': range(df['Year'].min(), df['Year'].max() + 1)})
df = full_years.merge(df, on='Year', how='left')

df['smokers_percentage_no'] = df['smokers_percentage_no'].interpolate(method='linear')
df['smokers_percentage_sw'] = df['smokers_percentage_sw'].interpolate(method='linear')
df['smokers_percentage_den'] = df['smokers_percentage_den'].interpolate(method='linear')
df['smokers_percentage_fin'] = df['smokers_percentage_fin'].interpolate(method='linear')

# Show first few rows
df = df.tail(27)
df

Year  smokers_percentage_no  smokers_percentage_sw  \
25  1998                   33.0                  18.00   
26  1999                   31.0                  17.75   
27  2000                   31.0                  17.50   
28  2001                   29.0                  17.00   
29  2002                   29.0                  16.50   
30  2003                   26.0                  15.75   
31  2004                   26.0                  15.00   
32  2005                   25.0                  14.50   
33  2006                   23.0                  14.00   
34  2007                   21.0                  13.80   
35  2008                   20.0                  13.00   
36  2009                   20.0                  12.75   
37  2010                   19.0                  12.50   
38  2011                   17.0                  12.15   
39  2012                   16.0                  11.80   
40  2013                   15.0                  11.40   
41  2014                   13.0                  11.00   
42  2015                   13.0                  10.75   
43  2016                   12.0                  10.50   
44  2017                   11.0                  10.25   
45  2018                   11.0                  10.00   
46  2019                    9.0                   9.90   
47  2020                    9.0                   9.80   
48  2021                    8.0                   9.70   
49  2022                    8.0                   9.50   
50  2023                    7.0                   8.80   
51  2024                    8.0                   8.80   

    smokers_percentage_den  smokers_percentage_fin  
25                    35.0                    18.5  
26                    34.0                    18.0  
27                    33.0                    18.0  
28                    32.0                    18.0  
29                    31.0                    17.0  
30                    30.0                    16.0  
31                    27.0                    16.0  
32                    28.0                    16.0  
33                    28.0                    15.5  
34                    28.0                    15.0  
35                    28.0                    14.5  
36                    24.0                    14.0  
37                    23.0                    13.5  
38                    22.0                    13.0  
39                    23.0                    12.0  
40                    22.0                    11.0  
41                    21.0                    14.0  
42                    22.0                    14.0  
43                    21.0                    13.0  
44                    22.0                    11.0  
45                    23.0                    12.0  
46                    20.0                    11.0  
47                    18.0                    11.0  
48                    19.0                    10.5  
49                    19.0                    10.0  
50                    18.0                    10.0  
51                    17.0                    10.0

In [265]:
%pip install pyfixest

Note: you may need to restart the kernel to use updated packages.


In [266]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pyfixest.estimation import feols
# from pyfixest.visualize import iplot

# Reshape to long format
df_long = df.melt(id_vars='Year', value_vars=['smokers_percentage_no', 'smokers_percentage_sw', 
                                              'smokers_percentage_den', 'smokers_percentage_fin'],
                  var_name='Country', value_name='Smokers')

# Clean up country names
df_long['Country'] = df_long['Country'].str.replace('smokers_percentage_', '')

# --- Event Study Setup ---
# 1. Define the ban year for each country.
ban_years = {
    'no': 2004,
    'sw': 2005,
    'den': 2007,
    'fin': 2005
}
df_long['BanYear'] = df_long['Country'].map(ban_years)

# 2. Calculate "event time"
df_long['EventTime'] = df_long['Year'] - df_long['BanYear']

# 3. Run the event study regression using fixest
# The formula uses `i(EventTime, ref=-1)` for the event study interaction.
# The pipe `|` is used to specify the fixed effects (Country and Year).
# `cluster='Country'` correctly handles the standard errors.
fit = feols('Smokers ~ i(EventTime, ref=-1) | Country + Year', 
            data=df_long)

# --- FIX: Specify clustered standard errors using the .vcov() method ---
# The standard errors are calculated after fitting the model.
fit.vcov({'CRV1': 'Country'})

print("--- Event Study Regression Results (using fixest) ---")
fit.summary()

# # 4. Visualize the results using fixest's iplot
# # This automatically creates a publication-quality event study plot.
# fig, ax = plt.subplots(figsize=(12, 7))
# iplot(fit, ax=ax)
# plt.title('Event Study: Effect of Smoking Bans on Smoking Rates')
# plt.xlabel('Years Relative to Ban Implementation (Event Time)')
# plt.ylabel('Change in Smoking Percentage Points')
# plt.grid(True, which='both', linestyle='--', linewidth=0.5)
# plt.show()

--- Event Study Regression Results (using fixest) ---
###

Estimation:  OLS
Dep. var.: Smokers, Fixed effects: Country+Year
Inference:  CRV1
Observations:  108

| Coefficient                                  |   Estimate |   Std. Error |   t value |   Pr(>|t|) |    2.5% |   97.5% |
|:---------------------------------------------|-----------:|-------------:|----------:|-----------:|--------:|--------:|
| C(EventTime, contr.treatment(base=-1))[T.-9] |     -7.953 |       17.851 |    -0.446 |      0.686 | -64.762 |  48.855 |
| C(EventTime, contr.treatment(base=-1))[T.-8] |     -5.047 |       14.838 |    -0.340 |      0.756 | -52.268 |  42.174 |
| C(EventTime, contr.treatment(base=-1))[T.-7] |    -10.378 |       15.834 |    -0.655 |      0.559 | -60.769 |  40.013 |
| C(EventTime, contr.treatment(base=-1))[T.-6] |     -6.057 |       11.958 |    -0.507 |      0.647 | -44.114 |  32.000 |
| C(EventTime, contr.treatment(base=-1))[T.-5] |     -5.729 |        7.942 |    -0.721 |      0.523 | -31.0

/Users/ulrikhaugland/data_science_venv/lib/python3.12/site-packages/pyfixest/estimation/feols_.py:2703: UserWarning: 
            1 variables dropped due to multicollinearity.
            The following variables are dropped: ['C(EventTime, contr.treatment(base=-1))[T.20]'].
            
  warnings.warn(


Testing the Parallel Trends Assumption (The Pre-Ban Period):

Look at the rows where EventTime is negative (e.g., [T.-9] to [T.-2]). These are the years before the smoking ban was implemented.
What to look for: You want the coefficients for these pre-ban years to be statistically insignificant. This means their p-values (Pr(>|t|)) should be large (typically > 0.10).
Your Result: For T.-9, the p-value is 0.686. For T.-2, it's 0.425. In fact, all of your pre-ban coefficients have large p-values. This is great news! It means there was no statistically significant difference in smoking trends between the countries before the bans were enacted. This provides strong evidence that the parallel trends assumption holds.

Estimating the Effect of the Ban (The Post-Ban Period):

Look at the rows where EventTime is zero or positive (e.g., [T.0], [T.1], etc.). These are the years during and after the ban.
What to look for: You are looking for statistically significant coefficients (p-values < 0.05 or < 0.10).
Your Result:
C(EventTime...)[T.0]: The coefficient is 1.443 with a p-value of 0.587. This means in the year of the ban, there was no immediate, statistically significant change in smoking rates.
C(EventTime...)[T.18]: The coefficient is 5.768 with a p-value of 0.094. This is significant at the 10% level. It suggests that 18 years after the ban, smoking rates were about 5.8 percentage points higher than they were the year before the ban, which is a very counter-intuitive result.
Overall Post-Ban Trend: None of your post-ban coefficients are significant at the 5% level, and the estimates are surprisingly positive. This suggests that, based on your data and model, the smoking bans did not have a statistically significant negative effect on smoking rates.

In [267]:
tobacco_price_no = pd.read_excel("tobacco_nordics/norway.xlsx", sheet_name="Monthly")
# Assume your DataFrame is called df_price
tobacco_price_no['observation_date'] = pd.to_datetime(tobacco_price_no['observation_date'])
tobacco_price_no['Year'] = tobacco_price_no['observation_date'].dt.year
tobacco_price_no['Month'] = tobacco_price_no['observation_date'].dt.month

# Filter for January only
tobacco_price_no = tobacco_price_no[tobacco_price_no['Month'] == 1].copy()
tobacco_price_no


observation_date  CP0220NOM086NEST  Year  Month
0         1996-01-01              30.9  1996      1
12        1997-01-01              34.8  1997      1
24        1998-01-01              39.7  1998      1
36        1999-01-01              41.4  1999      1
48        2000-01-01              45.3  2000      1
60        2001-01-01              47.8  2001      1
72        2002-01-01              49.4  2002      1
84        2003-01-01              50.8  2003      1
96        2004-01-01              60.4  2004      1
108       2005-01-01              62.7  2005      1
120       2006-01-01              64.2  2006      1
132       2007-01-01              65.8  2007      1
144       2008-01-01              68.7  2008      1
156       2009-01-01              72.1  2009      1
168       2010-01-01              77.1  2010      1
180       2011-01-01              83.3  2011      1
192       2012-01-01              86.3  2012      1
204       2013-01-01              89.6  2013      1
216       2014-01-01              93.8  2014      1
228       2015-01-01              98.2  2015      1
240       2016-01-01             102.5  2016      1
252       2017-01-01             107.2  2017      1
264       2018-01-01             111.6  2018      1
276       2019-01-01             114.4  2019      1
288       2020-01-01             118.3  2020      1
300       2021-01-01             115.6  2021      1
312       2022-01-01             118.9  2022      1
324       2023-01-01             125.9  2023      1
336       2024-01-01             131.9  2024      1
348       2025-01-01             142.3  2025      1

In [268]:

# Keep only one row per year (the January value)
tobacco_price_no = tobacco_price_no[['Year', 'CP0220NOM086NEST', 'observation_date']].reset_index(drop=True)
tobacco_price_no = tobacco_price_no.rename(columns={"CP0220NOM086NEST": "tobacco_price_no"}).drop(columns=["observation_date"])



tobacco_price_no

Year  tobacco_price_no
0   1996              30.9
1   1997              34.8
2   1998              39.7
3   1999              41.4
4   2000              45.3
5   2001              47.8
6   2002              49.4
7   2003              50.8
8   2004              60.4
9   2005              62.7
10  2006              64.2
11  2007              65.8
12  2008              68.7
13  2009              72.1
14  2010              77.1
15  2011              83.3
16  2012              86.3
17  2013              89.6
18  2014              93.8
19  2015              98.2
20  2016             102.5
21  2017             107.2
22  2018             111.6
23  2019             114.4
24  2020             118.3
25  2021             115.6
26  2022             118.9
27  2023             125.9
28  2024             131.9
29  2025             142.3

In [269]:
tobacco_price_fin = pd.read_excel("tobacco_nordics/finland.xlsx", sheet_name="Monthly")
tobacco_price_fin


observation_date  CP0220FIM086NEST
0         1996-01-01             53.88
1         1996-02-01             53.88
2         1996-03-01             53.88
3         1996-04-01             53.88
4         1996-05-01             53.88
..               ...               ...
352       2025-05-01            203.61
353       2025-06-01            204.58
354       2025-07-01            206.96
355       2025-08-01            212.23
356       2025-09-01            212.23

[357 rows x 2 columns]

In [270]:
# Assume your DataFrame is called df_price
tobacco_price_fin['observation_date'] = pd.to_datetime(tobacco_price_fin['observation_date'])
tobacco_price_fin['Year'] = tobacco_price_fin['observation_date'].dt.year
tobacco_price_fin['Month'] = tobacco_price_fin['observation_date'].dt.month

# Filter for January only
tobacco_price_fin = tobacco_price_fin[tobacco_price_fin['Month'] == 1].copy()

# Keep only one row per year (the January value)
tobacco_price_fin = tobacco_price_fin[['Year', 'CP0220FIM086NEST', 'observation_date']].reset_index(drop=True)
tobacco_price_fin = tobacco_price_fin.rename(columns={"CP0220FIM086NEST": "tobacco_price_fin"}).drop(columns=["observation_date"])



tobacco_price_fin

Year  tobacco_price_fin
0   1996              53.88
1   1997              55.44
2   1998              57.01
3   1999              57.01
4   2000              58.73
5   2001              61.21
6   2002              61.81
7   2003              61.80
8   2004              62.21
9   2005              61.37
10  2006              63.90
11  2007              66.30
12  2008              67.60
13  2009              74.77
14  2010              78.18
15  2011              79.42
16  2012              81.02
17  2013              89.23
18  2014              90.38
19  2015              99.05
20  2016             101.36
21  2017             112.08
22  2018             122.61
23  2019             131.78
24  2020             140.89
25  2021             152.74
26  2022             162.13
27  2023             175.32
28  2024             185.79
29  2025             203.58

In [271]:
tobacco_price_sw = pd.read_excel("tobacco_nordics/sweden.xlsx", sheet_name="Monthly")
tobacco_price_sw


observation_date  CP0220SEM086NEST
0         1996-01-01             40.21
1         1996-02-01             40.24
2         1996-03-01             40.18
3         1996-04-01             40.18
4         1996-05-01             40.11
..               ...               ...
352       2025-05-01            133.19
353       2025-06-01            133.50
354       2025-07-01            134.35
355       2025-08-01            134.85
356       2025-09-01            135.32

[357 rows x 2 columns]

In [272]:
# Assume your DataFrame is called df_price
tobacco_price_sw['observation_date'] = pd.to_datetime(tobacco_price_sw['observation_date'])
tobacco_price_sw['Year'] = tobacco_price_sw['observation_date'].dt.year
tobacco_price_sw['Month'] = tobacco_price_sw['observation_date'].dt.month

# Filter for January only
tobacco_price_sw = tobacco_price_sw[tobacco_price_sw['Month'] == 1].copy()

# Keep only one row per year (the January value)
tobacco_price_sw = tobacco_price_sw[['Year', 'CP0220SEM086NEST', 'observation_date']].reset_index(drop=True)
tobacco_price_sw = tobacco_price_sw.rename(columns={"CP0220SEM086NEST": "tobacco_price_sw"}).drop(columns=["observation_date"])



tobacco_price_sw

Year  tobacco_price_sw
0   1996             40.21
1   1997             42.13
2   1998             56.85
3   1999             47.66
4   2000             48.08
5   2001             49.39
6   2002             50.54
7   2003             51.94
8   2004             53.35
9   2005             53.91
10  2006             55.59
11  2007             63.78
12  2008             73.94
13  2009             76.87
14  2010             80.31
15  2011             82.27
16  2012             86.16
17  2013             93.23
18  2014             95.08
19  2015             97.07
20  2016            101.24
21  2017            103.25
22  2018            104.84
23  2019            106.91
24  2020            109.20
25  2021            112.36
26  2022            114.27
27  2023            119.17
28  2024            126.79
29  2025            129.31

In [273]:
tobacco_price_den = pd.read_excel("tobacco_nordics/denmark.xlsx", sheet_name="Monthly")
tobacco_price_den


observation_date  CP0220DKM086NEST
0         1996-01-01              59.4
1         1996-02-01              59.5
2         1996-03-01              60.4
3         1996-04-01              60.5
4         1996-05-01              60.5
..               ...               ...
352       2025-05-01             159.9
353       2025-06-01             160.0
354       2025-07-01             160.4
355       2025-08-01             162.6
356       2025-09-01             162.9

[357 rows x 2 columns]

In [274]:
# Assume your DataFrame is called df_price
tobacco_price_den['observation_date'] = pd.to_datetime(tobacco_price_den['observation_date'])
tobacco_price_den['Year'] = tobacco_price_den['observation_date'].dt.year
tobacco_price_den['Month'] = tobacco_price_den['observation_date'].dt.month

# Filter for January only
tobacco_price_den = tobacco_price_den[tobacco_price_den['Month'] == 1].copy()

# Keep only one row per year (the January value)
tobacco_price_den = tobacco_price_den[['Year', 'CP0220DKM086NEST', 'observation_date']].reset_index(drop=True)
tobacco_price_den = tobacco_price_den.rename(columns={"CP0220DKM086NEST": "tobacco_price_den"}).drop(columns=["observation_date"])



tobacco_price_den

Year  tobacco_price_den
0   1996               59.4
1   1997               60.5
2   1998               61.6
3   1999               62.7
4   2000               63.7
5   2001               66.8
6   2002               68.4
7   2003               69.6
8   2004               63.2
9   2005               67.1
10  2006               67.2
11  2007               68.0
12  2008               69.4
13  2009               70.6
14  2010               78.8
15  2011               85.8
16  2012               88.8
17  2013               96.2
18  2014               99.8
19  2015               99.9
20  2016              100.1
21  2017              100.4
22  2018              100.7
23  2019              101.5
24  2020              103.6
25  2021              141.2
26  2022              138.3
27  2023              152.8
28  2024              155.2
29  2025              157.6

In [275]:

tobacco_price_sw['Year'] = pd.to_numeric(tobacco_price_sw['Year'], errors='coerce').astype(int)
tobacco_price_no['Year'] = pd.to_numeric(tobacco_price_no['Year'], errors='coerce').astype(int)
tobacco_price_den['Year'] = pd.to_numeric(tobacco_price_den['Year'], errors='coerce').astype(int)
tobacco_price_fin['Year'] = pd.to_numeric(tobacco_price_fin['Year'], errors='coerce').astype(int)

In [279]:
# --- 1. Create a single WIDE DataFrame for prices ---
df_price_wide = tobacco_price_sw.merge(
    tobacco_price_fin, on="Year", how="outer"
).merge(
    tobacco_price_no, on="Year", how="outer"
).merge(
    tobacco_price_den, on="Year", how="outer"
)

# --- 2. Melt the price data from WIDE to LONG format ---
df_price_long = df_price_wide.melt(
    id_vars='Year',
    value_vars=['tobacco_price_sw', 'tobacco_price_fin', 'tobacco_price_no', 'tobacco_price_den'],
    var_name='Country',
    value_name='TobaccoPrice'
)

# --- 3. Clean the 'Country' names to match df_long ---
df_price_long['Country'] = df_price_long['Country'].str.replace('tobacco_price_', '')

# --- 4. Merge the price data into the main analysis DataFrame ---
df_final = pd.merge(df_long, df_price_long, on=['Year', 'Country'], how='left')

# --- 5. Run the Event Study with TobaccoPrice as a control variable ---
# --- FIX: Remove 'cluster' from feols() and use the .vcov() method instead ---
fit_with_controls = feols('Smokers ~ i(EventTime, ref=-1) + TobaccoPrice | Country + Year', 
                          data=df_final)

fit_with_controls.vcov({'CRV1': 'Country'})

print("--- Event Study Regression Results (with Tobacco Price Control) ---")
fit_with_controls.summary()



--- Event Study Regression Results (with Tobacco Price Control) ---
###

Estimation:  OLS
Dep. var.: Smokers, Fixed effects: Country+Year
Inference:  CRV1
Observations:  108

| Coefficient                                  |   Estimate |   Std. Error |   t value |   Pr(>|t|) |    2.5% |   97.5% |
|:---------------------------------------------|-----------:|-------------:|----------:|-----------:|--------:|--------:|
| C(EventTime, contr.treatment(base=-1))[T.-9] |     -9.381 |       19.074 |    -0.492 |      0.657 | -70.084 |  51.322 |
| C(EventTime, contr.treatment(base=-1))[T.-8] |     -6.320 |       15.883 |    -0.398 |      0.717 | -56.867 |  44.227 |
| C(EventTime, contr.treatment(base=-1))[T.-7] |    -11.598 |       16.776 |    -0.691 |      0.539 | -64.987 |  41.791 |
| C(EventTime, contr.treatment(base=-1))[T.-6] |     -6.895 |       12.544 |    -0.550 |      0.621 | -46.816 |  33.026 |
| C(EventTime, contr.treatment(base=-1))[T.-5] |     -6.402 |        8.353 |    -0.766 |     

/Users/ulrikhaugland/data_science_venv/lib/python3.12/site-packages/pyfixest/estimation/feols_.py:2703: UserWarning: 
            1 variables dropped due to multicollinearity.
            The following variables are dropped: ['C(EventTime, contr.treatment(base=-1))[T.20]'].
            
  warnings.warn(
